sw/device/tests/dif/dif_pinmux_test.cc

```diff
@@ -0,0 +1,378 @@
+      EXPECT_WRITE32(reg, 0x20820820);
+    }
+
+    mmio_region_t base_addr = dev().region();
+    EXPECT_EQ(dif_pinmux_init(base_addr, &dif_pinmux_),
+              kDifPinmuxInitErrorNone);
+
+    EXPECT_TRUE(dif_pinmux_.write_enabled);
+  }
+
+  dif_pinmux_t dif_pinmux_;
+};
+
+// dif_pinmux_init() tests.
+
+TEST_F(DifPinmuxTest, InitFailureInvalidArgumentPinmux) {
+  mmio_region_t base_addr = dev().region();
+  EXPECT_EQ(dif_pinmux_init(base_addr, nullptr),
+            kDifPinmuxInitErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, InitFailureRegistersLocked) {
+  EXPECT_READ32(PINMUX_REGEN_REG_OFFSET, (uint32_t)0);
+
+  mmio_region_t base_addr = dev().region();
+  EXPECT_EQ(dif_pinmux_init(base_addr, &dif_pinmux_),
+            kDifPinmuxInitErrorRegistersLocked);
+
+  EXPECT_FALSE(dif_pinmux_.write_enabled);
+}
+
+TEST_F(DifPinmuxTest, InitSuccess) { DeviceInit(); }
+
+// dif_pinmux_registers_lock() tests.
+
+TEST_F(DifPinmuxTest, RegistersLockFailureInvalidArgumentPinmux) {
+  EXPECT_EQ(dif_pinmux_registers_lock(nullptr),
+            kDifPinmuxInitErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, RegistersLockSuccess) {
+  DeviceInit();
+
+  // TODO - when fixed, drop the explicit cast.
+  EXPECT_WRITE32(PINMUX_REGEN_REG_OFFSET, (uint32_t)0);
+
+  EXPECT_EQ(dif_pinmux_registers_lock(&dif_pinmux_),
+            kDifPinmuxRegistersLockErrorNone);
+
+  EXPECT_FALSE(dif_pinmux_.write_enabled);
+}
+
+// dif_pinmux_insel_set() tests.
+
+TEST_F(DifPinmuxTest, InselSetFailureInvalidArgumentPinmux) {
+  EXPECT_EQ(dif_pinmux_insel_set(nullptr, kDifPinmuxPeripheralInGpio0,
+                                 kDifPinmuxInselTieToZero),
+            kDifPinmuxInselSetErrorInvalidArgument);
+}
+
+TEST_F(DifPinmuxTest, InselSetFailureBadMioSelect) {
+  DeviceInit();
+
+  int less_than_min = (int)kDifPinmuxInselTieToZero - 1;
+  EXPECT_EQ(dif_pinmux_insel_set(&dif_pinmux_, kDifPinmuxPeripheralInGpio0,
+                                 (dif_pinmux_insel_t)less_than_min),
+            kDifPinmuxInselSetErrorBadMioSelect);
+
+  int more_than_max = (int)kDifPinmuxInselPadLast + 1;
+  EXPECT_EQ(dif_pinmux_insel_set(&dif_pinmux_, kDifPinmuxPeripheralInGpio0,
+                                 (dif_pinmux_insel_t)more_than_max),
+            kDifPinmuxInselSetErrorBadMioSelect);
+}
+
+TEST_F(DifPinmuxTest, InselSetFailureRegistersLocked) {
+  DeviceInit();
+
+  dif_pinmux_.write_enabled = false;
+  EXPECT_EQ(dif_pinmux_insel_set(&dif_pinmux_, kDifPinmuxPeripheralInGpio0,
+                                 kDifPinmuxInselTieToZero),
+            kDifPinmuxInselSetErrorRegistersLocked);
+
+  EXPECT_FALSE(dif_pinmux_.write_enabled);
+}
+
+TEST_F(DifPinmuxTest, InselSetSuccessFirstRegister) {
+  DeviceInit();
+
+  // Set the first insel field value to tie zero.
+  EXPECT_MASK32(
+      PINMUX_PERIPH_INSEL0_REG_OFFSET,
+      {
+          {PINMUX_PERIPH_INSEL0_IN0_OFFSET, PINMUX_PERIPH_INSEL0_IN0_MASK,
+           (uint32_t)kDifPinmuxInselTieToZero},
```

Comment:
I can never remember... does C++ really not define an implicit conversion for this? Yuck.

Could always just do
```
#ifdef __cplusplus
#define ENUM32(foo) enum foo : uint32_t
#else
#define ENUM32(foo) enum foo
#endif
```
but that's kind of a bad idea.

Sigh, C++.

Reply:
I am not sure what my opinion on this is, I think normally I prefer explicit casts.